### ISSUE TITLE
Gitable.py tweaks

### ISSUE THREAD

**Issue by macluo** [CSC510-2015-Axitron/project2#2]
The original gitable.py only grabs label and unlabel two actions. I thought that was rather limited so here is the revised gitable 70e4bd874c9e3c76196f83503db791760b303bb6. Hope it is useful.

Now gitable-v2.py should also grab multiple labels, assignee, and additional actions. Next version will anonymize user IDs.

**Comment by macluo** [comment]
Anonymized version is now included 86f983807253d971c2b28314da7f23fdf9d4a46b

**Comment by macluo** [comment]
Minor fix on labels and added anonymizer flag to v2.1 b6c36f129b8cf4fcfe606413e702a8ecae15050c & f3435b807bc1462b368db1f36aa80b6d3b3c66c3

**Comment by duh102** [comment]
I included anonymization functions into the sql version I created, didn't merge in the assignee/other actions bits.

**Comment by macluo** [comment]
:+1:

**Comment by duh102** [comment]
Merged in retrieval of other events (in fact, all events tracked). We may want to customize further how the events output their 'what' attributes, currently only labeling, milestones, and assignments give real data, the others just copy over their action. For now, users are anonymized both in their 'what' and 'user' attributes, and milestones are converted to an index into the milestones table both in 'what' and 'milestone'.
Ex:

```
action : closed,        what : closed,  when : 1428349862.0,    user : macluo
action : labeled,       what : 4. feature complete,     when : 1428026300.0,    user : duh102
action : assigned,      what : macluo,  when : 1428325992.0,    user : macluo
action : milestoned,    what : v0.6 Cleanup,    when : 1427838875.0,    user : ktspence,        milestone : v0.6 Cleanup
```

The code to do this conditional formatting is a bit messy (just a nested ternary), but it works for now. Might want to fix it at some point.

I also had to eliminate duplicates, seems that occasionally a de/milestoning event would be placed twice into the log, hence why you might see some output that looks like

```
duplicates: [7, 1424538797.0, u'milestoned', 2, 'CSC510-2015-Axitron/maze/user1', 2]
```

**Comment by duh102** [comment]
Small fix in 03667fc3a54fc9ff22b8b1b529c299ea1bf872c2 adds in the github event id for each tracked event, in case you want to enforce strict event ordering (ex: when timestamps are identical).

**Comment by duh102** [comment]
For an example of a "bad smell" detector, consider the following SQL:

``` sql
select cl.issueID, (cl.time - op.time) as timeOpen from event cl, (select issueID, min(time) as time from event group by issueID) op where cl.action == 'closed' AND cl.issueID == op.issueID;
```

This lists the amount of time (in seconds) from when an issue is opened (has some event) to when the issue is marked as closed, for each issue. If we do statistical analysis over all of these for every project we do, then we can find out what a typical issue should take (for a project of our scope and time limit) and then detect which issues are taking "too long".

**Comment by macluo** [comment]
Just a thought. We could also exclude those issues marked as "Won't fix" as their time may be very different than those have been worked on. Hopefully the projects we will be looking at have all the proper labels in place.

Another question, is it possible to run gitable-sql.py without updating to sql? I just thought sometime we might want to take a quick look at the events & labels but do not want the data sent to the database again.

**Comment by duh102** [comment]
Yes, just change the db in the config to `:memory:` instead of a filename, that will pull all the events and insert them into an in-memory database that will be deallocated as soon as the program ends.

**Comment by macluo** [comment]
cool!

**Comment by duh102** [comment]
I should mention that you don't have to use SQL to query the database if you don't want, you can always grab all the items and do manual queries using some other program (or somewhere in gitable-sql.py).

**Comment by duh102** [comment]
Just pushed a big change in 01c44bc52d0903ea5d713a1c6ed287aa51c05863, turns out that milestones have IDs, and events do not have those IDs when milestoning or demilestoning, only titles, and those titles may change. The milestone IDs _are_, however, stored in the event's milestone field, if it exists. I also changed the way you use gitable-sql.py, and deleted the other versions of it sitting in the same directory (felt a bit messy).
New usage is similar to previous version in that you need token in config file, but you run with the name of the repo (And optionally the name of the resulting db), as below:

```
usage: gitable-sql.py [-h] [-db DATABASE] repo

Process GitHub issue records and record to SQLite database

positional arguments:
  repo                  repo to process

optional arguments:
  -h, --help            show this help message and exit
  -db DATABASE, --database DATABASE
                        specify db filename, default (repo).db
```

This makes it easier to automate pulling records, just execute run.sh and it'll pull all records for all repos in the class.
Data formats are as follows for each table.

#### Issues

| id | name |
| --- | --- |
| 1 | User Input (moving smily via keyboard) |

#### Milestones

| id | title | description | created_at | due_at | closed_at | user | identifier |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | v0.1 | (desc) | 1424538641 | 1425013200 | 1425168047 | CSC510-2015-Axitron/maze/user1 | 989453 |

#### Events

| issueID | time | action | label | user | milestone | identifier |
| --- | --- | --- | --- | --- | --- | --- |
| 1 | 1423616709 | closed | closed | CSC510-2015-Axitron/maze/user1 |  | 233661242 |

**Comment by duh102** [comment]
Probably should update what all the tables look like now, since they've changed a bit.

#### Milestone

| id | title | description | created_at | due_at | closed_at | user | identifier |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | v0.1 | (text) | 1424538641 | 1425013200 | 1425168047 | group9/user1 | 989453 |

#### Issue

| id | name |
| --- | --- |
| 1 | User Input (moving smily via keyboard) |

#### Event

| issueID | time | action | label | user | milestone | identifier |
| --- | --- | --- | --- | --- | --- | --- |
| 1 | 1423616709 | closed | closed | group9/user1 |  | 233661242 |

#### Comment

| issueID | user | createtime | updatetime | text | identifier |
| --- | --- | --- | --- | --- | --- |
| 1 | group9/user1 | 1423608247 | 1423608247 | (text) | 73800452 |

#### Commits

| id | time | sha | user | message |
| --- | --- | --- | --- | --- |
| 1 | 1428349546 | d924a137ebf530fcd56c73980c9fcfbf6de69cdd | group9/user3 | (text) |
